hubblestack/extmods/fdg/command_line_parser.py

```diff
@@ -0,0 +1,193 @@
+1. ([\", \']) : () brackets signify a group. This is the second group in this regex.
+                    [\",\'] inside the () matches either a double-quote or a single-quote.
+2. .*? : matches any character lazily.
+3. \\2 : matches the second group of the regex.
+
+regex2 :
+    "(?<=(\s|{))-{0,2}\"{0,1}\'{0,1}" + key_alias + "\"{0,1}\'{0,1}\s*" + delimiter + "\s*.+?(?=(\s|$|}))"
+The starting regex is similar to the regex_pattern.
+
+1. (?=(\s|$)) : matches that the end of the pattern must contain either <space> or EOL
+"""
+
+import logging
+import re
+
+log = logging.getLogger(__name__)
+open_bracket_list = ["[","{","("]
+close_bracket_list = ["]","}",")"]
+
+
+def parse_cmdline(params='', chained=None, chained_status=None):
+    try:
+        if not params:
+            ret = {'Failure': 'invalid input, no params provided'}
+            return False, ret
+
+        if not params.get('key_aliases'):
+            log.error("No key_aliases provided in params to parse_cmdline, returning False")
+            return False, ''
+        else:
+            key_aliases = params.get('key_aliases')
+
+        if not chained:
+            log.error("No chained value provided to parse_cmdline function, returning False")
+            return False, ''
+        elif not chained.get('cmdline'):
+            log.error("cmdline not provided to parse_cmdline function, returning False")
+            return False, ''
+        else:
+            command_line = chained.get('cmdline')
+
+        if not params.get('delimiter') or len(params.get('delimiter')) > 1:
+            log.info("Either length of delimiter exceeds 1 or no delimiter provided with "
+                     "command line '%s'", command_line)
+            return False, ''
+        else:
+            delimiter = params.get('delimiter')
+
+        log.debug("value of command_line is %s", command_line)
+        log.debug("value of key_aliases is %s", key_aliases)
+        log.debug("value of delimiter is '%s'", delimiter)
+        ret_match_list = []
+
+        for key_alias in key_aliases:
+            log.debug("looping with key_alias %s", key_alias)
+            key_alias = key_alias.lstrip("-")
+
+            if key_alias not in command_line:
+                log.info("key_alias %s not found in command line %s", key_alias, command_line)
+                continue
+
+            regex_list = []
+            regex_pattern = "(?<=(\s|{))-{0,2}\"{0,1}\'{0,1}" + key_alias + "\"{0,1}\'{0,1}\s*" + delimiter + "\s*openingbracket.*closingbracket(?=(\s|$))"
+            braces_list = [('\[','\]'), ('\{','\}'), ('\(','\)')]
+            for item in braces_list:
+                regex = re.sub("openingbracket", item[0], regex_pattern)
+                regex = re.sub("closingbracket", item[1], regex)
+                regex_list.append(regex)
+            regex1 = "(?<=(\s|{))-{0,2}\"{0,1}\'{0,1}" + key_alias + "\"{0,1}\'{0,1}\s*" + delimiter + "\s*([\",\']).*?\\2"
```

Comment:
String concatenation by '+' is bad idea in almost every language, especially when we are appending many strings.

Can use join or f-string.

Reply:
Done. However can you please tell why is '+' not a good option?

Reply:
Multiple appends with '+' creates new string object for each two appends. Since string is mutable. Also, if appending more than 2 strings with '+' causes complexity of O(n^2)

Also see this PEP recommendation: https://www.python.org/dev/peps/pep-0008/#programming-recommendations